### ISSUE TITLE
'Module has no attribute "xxx"' when mixing `py_binary` deps and `__init__` reexports

### ISSUE THREAD

**Issue by jimmyt857** [theoremlp/rules_mypy#74]
Thanks for the great work on this project! We've adopted it throughout parts of our monorepo and we really like it.

While trying to expand our usage of `rules_mypy`, I've run into an issue which I believe is a bug. I was able to produce a minimal example, which seems to have something to do with `__init__.py` files that depend on `py_binary` targets (rather than `py_library`).

# MRE

`example/BUILD.bazel`
```
py_binary(
    name = "main",
    srcs = ["main.py"],
    deps = ["//example/foo:__init__"],
)
```

`example/main.py`
```python
import example.foo as foo

if __name__ == "__main__":
    print(foo.do_something())
```

`example/foo/BUILD.bazel`
```
py_library(
    name = "__init__",
    srcs = ["__init__.py"],
    deps = [":a"],
)

py_binary(
    name = "a",
    srcs = ["a.py"],
)
```

`example/foo/__init__.py`
```python
from example.foo.a import do_something

__all__ = ["do_something"]
```

`example/foo/a.py`
```python
def do_something():
    return "hello"


if __name__ == "__main__":
    print(do_something())
```

## Result

When I run `bazel build //example/...` I see the following error:

```bash
example/main.py:4: error: Module has no attribute "do_something"  [attr-defined]
```

Outside of bazel, mypy does not complain about this.

**Comment by arrdem** [comment]
https://github.com/aspect-build/bazel-examples/compare/arrdem/mypy_74_repro At least in my testing this didn't repro, not sure what other details there may be here.

**Comment by arrdem** [comment]
After some more testing I'm pretty sure there's a `rules_mypy` bug here. The `MYPYPATH` that gets generated in this sort of case is extremely suspicious, the sandbox commands without the computed `MYPYPATH` set pass. There may also be interaction(s) with `aspect_rules_py` that are relevant here as well.

An example of the odd values; yaml format.

```
DEBUG: /private/var/tmp/_bazel_arrdem/bb363081b2f9ca1f305bb3de065f0f09/external/rules_mypy+/mypy/private/mypy.bzl:155:10: ---
name: @@//py_mypy/cli:cli
types:

pypi_dirs:
- bazel-out/darwin_arm64-fastbuild/bin/py_mypy/data_models
- bazel-out/darwin_arm64-fastbuild/bin/py_mypy/data_persistence
- bazel-out/darwin_arm64-fastbuild/bin/py_mypy/inventory_management/src
- bazel-out/darwin_arm64-fastbuild/bin/py_mypy/order_processing
external_deps:
- external/rules_python++pip+pip_311_click/site-packages
imports_dirs:
- py_mypy/data_models
- py_mypy/data_persistence
- py_mypy/inventory_management/src
- py_mypy/order_processing
- site-packages
generated_dirs:
- bazel-out/darwin_arm64-fastbuild-ST-dd1656cbd1bf/bin
- bazel-out/darwin_arm64-fastbuild/bin
generated_imports_dirs:
- bazel-out/darwin_arm64-fastbuild-ST-dd1656cbd1bf/bin/py_mypy/data_models
- bazel-out/darwin_arm64-fastbuild-ST-dd1656cbd1bf/bin/py_mypy/data_persistence
- bazel-out/darwin_arm64-fastbuild-ST-dd1656cbd1bf/bin/py_mypy/inventory_management/src
- bazel-out/darwin_arm64-fastbuild-ST-dd1656cbd1bf/bin/py_mypy/order_processing
- bazel-out/darwin_arm64-fastbuild-ST-dd1656cbd1bf/bin/site-packages
- bazel-out/darwin_arm64-fastbuild/bin/py_mypy/data_models
- bazel-out/darwin_arm64-fastbuild/bin/py_mypy/data_persistence
- bazel-out/darwin_arm64-fastbuild/bin/py_mypy/inventory_management/src
- bazel-out/darwin_arm64-fastbuild/bin/py_mypy/order_processing
- bazel-out/darwin_arm64-fastbuild/bin/site-packages
```
